operator/v1/types_network.go

```diff
@@ -562,6 +562,31 @@ type IPsecConfig struct {
 	// +kubebuilder:validation:Enum=Disabled;External;Full
 	// +optional
 	Mode IPsecMode `json:"mode,omitempty"`
+
+	// full defines configuration parameters for the IPsec `Full` mode.
+	// This will be effective only when mode is configured with `Full`,
+	// otherwise it will be ignored.
+	// +optional
+	Full IPsecFullModeConfig `json:"full,omitempty"`
+}
+
+type Encapsulation string
+
+const (
+	// DisableEncapsulation disables RFC-3948 encapsulation on the IPsec tunnels
+	// configured by OVN and OpenvSwitch.
+	EnableEncapsulation = "yes"
+	// DisableEncapsulation disables RFC-3948 encapsulation on the IPsec tunnels
+	// configured by OVN and OpenvSwitch.
+	DisableEncapsulation = "no"
+)
+
+type IPsecFullModeConfig struct {
+	// encapsulation option to enabled or disable rfc3949 encapsulation of IKE and IPsec
+	// encryption packets
+	// +kubebuilder:validation:Enum:=yes;no;""
```

Comment:
Yes and No are as bad as true and false. What does this actually do for an end user? What kind of encapsulation is this? Think about this from an end user perspective.

What difference does it make to me, as an end user, if I do, or do not have this field set to yes or no?

Reply:
This option is going to set into OVN/OVS something like this: https://docs.openvswitch.org/en/latest/tutorials/ipsec/#custom-options and finally going to be set as ipsec config option for an ipsec tunnel.


```
encapsulation

In some cases, for example when ESP packets are filtered or when a broken IPsec peer does not properly recognise NAT,
it can be useful to force RFC-3948 encapsulation. In other cases, where IKE is NAT'ed but ESP packets can or should
flow without encapsulation, it can be useful to ignore the NAT-Traversal auto-detection. encapsulation=yes forces
the NAT detection code to lie and tell the remote peer that RFC-3948 encapsulation (ESP in port 4500 packets) is
required. encapsulation=no ignores the NAT detection causing ESP packets to send send without encapsulation.
The default value of encapsulation=auto follows the regular outcome of the NAT auto-detection code performed
in IKE. This option replaced the obsoleted forceencaps option.
```
https://libreswan.org/man/ipsec.conf.5.html

So I don't see any other values to represent this. Please suggest.

Reply:
@JoelSpeed here is how implementation looks like for this new API: https://github.com/openshift/cluster-network-operator/pull/2573.

Reply:
Seems to me like there are three modes if I read this correctly? Force, Auto, Disable? With the default being Auto?

Could you briefly explain to me, what actually is encapsulation in this context? Does AH or ESP, transport or tunnel come into this in any way? https://support.huawei.com/enterprise/en/doc/EDOC1100301600/c5d5a9b/understanding-ipsec-encapsulation-modes

I'm still trying to understand how we explain this to an end user in terms that they can easily understand

Reply:
if ESP (IP Protocol 50)  is not supported this uses UDP encapsulation.  This is required for IBMCloud for example.

Reply:
If your network or firewall blocks less common IP protocols like IP Protocol 50, this setting allows the use of standard UDP packets on ports 500 and 4500 instead.

Reply:
IBMCloud IPI already uses UDP encap, it's just hardcoded.  This supports `platform: none` or other edge cases.

We should probably add UDP port 4500 somewhere in descriptions. 

Other UPI Docs already indicate UDP port 4500 is required, so we should be covered.
e.g. AWS UPI:
| Name                          | Description            | Protocol | Port |
|-------------------------------|------------------------|----------|------|
| MasterIngressIpsecNat         | IPsec NAT-T packets    | udp      | 4500 |
| MasterIngressWorkerIpsecNat   | IPsec NAT-T packets    | udp      | 4500 |

Reply:
Thanks @rbbratta about firewall use case, The NAT-T is very much needed there.

@JoelSpeed The IPsec tunnel is configured between OCP nodes in tunnel mode on the Geneve overlay traffic which carries pod payload traffic. Once IPsec SAs are established between nodes, then Gevenve traffic is encrypted with ESP header. So NAT-T (UDP) encapsulation happens on ESP payload so that it can pass firewalls and also compatible with intermediate NAT devices.

updated with 3 encapsulation modes (yes, no and auto) now.

Reply:
Can we please avoid the use of `yes` and `no` as enum values, they are as bad as true and false. Enum values should be descriptive where possible, I still think Force and Disable are better

Think of it this way:
* Yes encapsulation
* No encapsulation
vs
* Force encapsulation
* Disable encapsulation

The latter two flow better in a sentence at a minimum

Also, enum values should be PascalCase

Reply:
yes @JoelSpeed , done.